### PR TITLE
Add error state when pipeline version is deleted

### PR DESCRIPTION
**Pull request by ppadti**
Closes: [RHOAIENG-6683](https://issues.redhat.com/browse/RHOAIENG-6683)

## Description
This PR aims to add error state in pipeline run details page when pipeline version is deleted.

Screenshot from 2024-07-03 21-14-05

Screenshot from 2024-07-02 12-43-39

Screenshot from 2024-07-03 21-14-12

## How Has This Been Tested?
1. create a pipeline
2. create a run for that.
3. delete pipeline version of that run.
4. From Experiments and run page open pipeline run details page.
5. It will show the error state on run graph and pipeline spec tab. Also in details tab, "No pipeline version" will be mentioned for pipeline version property.

## Test Impact
Added cypress test.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by ppadti** [comment]
cc @yannnz

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx [hunk 1]
```diff
@@ -69,7 +69,7 @@ const PipelineRunDetails: PipelineCoreDetailsPageComponent = ({ breadcrumbPath,
   );
 
   const loaded = runLoaded && (versionLoaded || !!run?.pipeline_spec);
```
Comment (dpanshug) [inline frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx:71]: Change the loading condition to runLoaded only and you can check the version error in another variable.
```suggestion
  const loaded = runLoaded;
  const isVersionError = !versionLoaded || !run?.pipeline_spec
```

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx [hunk 2]
```diff
@@ -164,6 +165,7 @@ const PipelineRunDetails: PipelineCoreDetailsPageComponent = ({ breadcrumbPath,
               graphContent={
                 <PipelineTopology
                   nodes={nodes}
+                  versionError={versionError}
                   selectedIds={selectedId ? [selectedId] : []}
```
Comment (dpanshug) [inline frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx:169]: You can pass the `isVersionError` to the PipelineTopology to render the Error state.

File: frontend/src/concepts/topology/PipelineTopology.tsx [hunk 3]
```diff
@@ -37,6 +49,30 @@ const PipelineTopology: React.FC<PipelineTopologyProps> = ({
     return undefined;
   }, [controller, onSelectionChange]);
 
+  if (versionError) {
```
Comment (dpanshug) [inline frontend/src/concepts/topology/PipelineTopology.tsx:52]: Get the `isVersionError` prop to check the version is loaded or not.

**Review by dpanshug — CHANGES_REQUESTED** [review]
ApplicationsPage is a global component, so rather than adding Pipeline specific props to it, you should handle the versionLoaded condition inside the PipelineRunDetails.
See the suggestions below to implement this.

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx [hunk 4]
```diff
@@ -84,7 +84,7 @@ const PipelineRunDetails: PipelineCoreDetailsPageComponent = ({ breadcrumbPath,
     );
   }
 
-  if (!loaded || (!executionsLoaded && !executionsError)) {
+  if ((!loaded && !versionError) || (!executionsLoaded && !executionsError)) {
```
Comment (dpanshug) [inline frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx:87]: use versionLoaded instead of versionError

**Review by dpanshug — CHANGES_REQUESTED** [review]
LGTM, just a small change

**Comment by dpanshug** [comment]
After this if we redirect to the runs of a deleted pipeline version, we see the loading state of the pipeline in the breadcrumb. How should we handle this? @Gkrumbach07 @yannnz 
Screenshot 2024-06-27 at 3 50 38 PM

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx (hunk 4 above)
Comment (ppadti) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx:87]: I think we don't need this check here. so removing it from here.

**Comment by yannnz** [comment]
@dpanshug I think this is a bug. We don't need to show pipeline name in the breadcrumb, since the path to this page is from Selected Experiment to this selected Run. I remeber @DaoDaoNoCode is working on fix it.

**Comment by yannnz** [comment]
Hi @ppadti , the Pipeline spec tab has been missed.

**Comment by ppadti** [comment]
> Hi @ppadti , the Pipeline spec tab has been missed.

@Gkrumbach07 to display Pipeline spec tab, we need to have version. https://github.com/opendatahub-io/odh-dashboard/blob/main/frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/PipelineRunDetails.tsx#L173
Is it okay if we don't display that when the pipeline version is deleted or you want me to change the condition there.

**Comment by yannnz** [comment]
@ppadti I would suggest to follow the same design pattern for pipeline graph: displaying error message as the statement.

**Comment by yannnz** [comment]
Hi @kaedward , could you help to review the error state for the Pipeline spec tab? In the mock, I just updated the "Run pipeline graph" to "Run pipeline spec".
Screenshot 2024-07-02 at 23 16 55

**Comment by kaedward** [comment]
@yannnz, I think I made a suggestion for the run graph message, but reading it back now, I don't think it's super clear because "error loading BLANK" implies that the info can be loaded if an action is taken. If the item has been deleted, the info will never be available to load. So what do you think of these alternate messages?

"**Pipeline run graph unavailable**
The pipeline version that this run graph belongs to has been deleted."

"**Pipeline spec unavailable**
The pipeline version that this pipeline spec belongs to has been deleted."

**Comment by yannnz** [comment]
Thanks @kaedward , sounds good to me. 
> @yannnz, I think I made a suggestion for the run graph message, but reading it back now, I don't think it's super clear because "error loading BLANK" implies that the info can be loaded if an action is taken. If the item has been deleted, the info will never be available to load. So what do you think of these alternate messages?
> 
> "**Pipeline run graph unavailable** The pipeline version that this run graph belongs to has been deleted."
> 
> "**Pipeline spec unavailable** The pipeline version that this pipeline spec belongs to has been deleted."

**Comment by ppadti** [comment]
> Thanks @kaedward , sounds good to me.
> 
> > @yannnz, I think I made a suggestion for the run graph message, but reading it back now, I don't think it's super clear because "error loading BLANK" implies that the info can be loaded if an action is taken. If the item has been deleted, the info will never be available to load. So what do you think of these alternate messages?
> > "**Pipeline run graph unavailable** The pipeline version that this run graph belongs to has been deleted."
> > "**Pipeline spec unavailable** The pipeline version that this pipeline spec belongs to has been deleted."

thanks @kaedward @yannnz

**Comment by yannnz** [comment]
Looks good to me.

**Review by dpanshug — APPROVED** [review]
/lgtm

**Pull request merged**